design-documents/database/refactor-customer-import.md

```diff
@@ -0,0 +1,269 @@
+    }
+    
+    $query = $this->_connection->select()->from(
+        'customer_entity',
+        ['entity_id', 'email']
+    )->where('CONCAT(email,website_id) in (?)', $customerAddresses);
+    
+    if ($this->getBehavior($rowData) == Import::BEHAVIOR_DELETE) {
+        $entitiesToDelete = $this->_connection->fetchCol($query, 'entity_id');
+    } elseif ($this->getBehavior($rowData) == Import::BEHAVIOR_ADD_UPDATE) {
+        $entitiesToUpdate = $this->_connection->fetchAll($query, 'email');
+        
+        /* should filter $validBunchData[$email] by $entitiesToUpdate and split them in two arrays $entitiesToUpdate and $entitiesToCreate*/
+    }
+```
+
+With these two arrays we can use the row by email to create the data to import. 
+
+To generate new ids I recommend to add functions:
+
+`Sequence`
+
+```php
+public function getNextValueBunch(int $count): array
+{
+    $currentValue = $this->getCurrentValue();
+    $nextvalue = $currentValue + $count;
+    $this->resource->getConnection($this->connectionName)
+        ->insert($this->resource->getTableName($this->sequenceTable), ['sequence_value' => $nextValue]);
+    return range($currentValue, $nextValue);
+}
+```
+
+ `EntityMetadata` 
+
+```php
+public function generateIdentifierBunch(int $count) : array
+{
+    $nextIdentifiers = [];
+    if ($this->sequence) {
+        $nextIdentifiers = $this->sequence->getNextValueBunch($count);
+    }
+    return $nextIdentifiers;
+}
+```
+
+**Advantage:** 
+
+* We can apply this approach to Address and Customer and import should improve performance and stability  
+
+
+# Concept 4 -  Stored Function to generate Sequences
+
+I use as base the following technical Article  https://www.percona.com/blog/2008/04/02/stored-function-to-generate-sequences/
+
+Query that I run before I started performance testing: 
+
+```mysql
+CREATE TABLE `sequence` (
+  `type` varchar(20) NOT NULL,
+  `value` int(10) unsigned NOT NULL,
+  PRIMARY KEY  (`type`)
+) ENGINE=MyISAM DEFAULT CHARSET=latin1;
+
+INSERT INTO sequence VALUES ('customer', 1);
+
+DELIMITER //
+CREATE FUNCTION sequence(seq_type char (20)) RETURNS int
+BEGIN
+ UPDATE sequence SET value=last_insert_id(value+1) WHERE type=seq_type;
+ RETURN last_insert_id();
+END
+//
+DELIMITER ;
+```
+
+Implementation of ` _getNextEntityId()` 
+
+```php
+protected function _getNextEntityId()
+{
+  return $this->_connection->query("select sequence('customer')")->fetchColumn();
+}
+```
+
+* Time: 3.92 sec
+* IO: 511ms
+* CPU: 3.4sec
+* Network: 1.15 mb
+* Sequel Queries: 538.5ms 1,084rq
+
+# Summary
+
+For all 4 concepts we need to have a migration from `customer_entity` to `sequence` table idea like 
+`migrateSequneceColumnData(customer_entity,entity_id)`. I think we will get the most benefit with the implementation of 
+Concept 3 because we are sending less queries to the database.  
+All concepts can be implemented in a backward compatible way because we only touch constructors and protected functions 
```

Comment:
For distributed systems, it will not be possible to scale this kind of solution, where for generating ids we are relying on the database itself; we need to control at the program level how we want to generate the IDs. Twitter was using similar sequence for MySQL using AUTO_INCREMENT; but they realized it was not possible to scale this approach.

I would suggest looking into Twitter Snowflake
https://github.com/twitter-archive/snowflake/tree/snowflake-2010

Another article around this subject is this
https://www.callicoder.com/distributed-unique-id-sequence-number-generator/

Although currently many of Magento customers running on single-node database, but in the cloud we would need to support large scale bulk operations in distributed style, in which case we would be able to allow multiple threads pushing the data; so an optimized service that can handle this scale would be better approach; which doesn't rely on hitting database for Ids.

Snowflake uses 8 byte unsigned integers, currently Magento using 4 byte unsigned integer, not-sure whether it will be BIC, it seems like a PATCH? according to https://devdocs.magento.com/guides/v2.3/extension-dev-guide/versioning/codebase-changes.html

But the idea is to centralize the ID generation and take it outside of database itself; the service should be similarly scalable as other services to support the load.

This service can provide Bulk IDs as well, and many implementations I have seen, uses Redis or similar Cache to store that Bunch of IDs, in order to avoid unnecessary calls to ID Generation Service as well. We may still be able to use Magento SequenceInterface, while underlying calling this new Distributed ID Service.